### ISSUE TITLE
Kyutai Pocket TTS inference

### ISSUE THREAD

**Issue by ekzhang** [ekzhang/jax-js#59]
I saw this new [TTS model](https://kyutai.org/pocket-tts-technical-report) online, and it is quite a small model (100M) parameters while seeming to have quite good quality! They have an autoregressive transformer architecture and a lot of work put into making it fast specifically for CPU inference in their case (likely because that's portable).

## Why do this

On the web browser though, `jax-js` being pretty darn portable as well, I think this is potentially a good use case for the framework because:

1. **It's useful.** The alternative is the [Web Audio API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API), which sounds very robotic / is not great for speech synthesis. If this TTS actually _sounds natural_, it could be packaged as a nice zero-dependency library that anyone can `npm install` and run to embed as local inference on a site.
2. **The model is complex.** I have been reading the model code and it's a bit nontrivial, there are multiple KV caches for two transformers (FlowLM and the transformer of Mimi), plus a flow matching model and real-time sampling.

See also: https://github.com/kyutai-labs/pocket-tts/issues/1

## Overview of architecture

I will give an overview of the inference architecture based on my reading today. (This is also notes for when I come back to work on this, maybe over the weekend, need to sleep today haha)

Config:
- Sample rate = 24 kHz (24000)
- Frame rate = 12.5 Hz (each frame is 80 ms of audio)
- Frame size = sample rate / frame rate = 1920
- `lsd_decode_steps = 1` - _the latent flow model is actually only sampled for one step! it simplifies inference._

**Input:** A text sample and an audio preconditioning (10 seconds of audio).

- **Voice embedding (can be precomputed) — should have shape `[12.5*t, 1024]`**
  - (Mimi encoding)
    - Resample audio preconditioning to 24 kHz `float[24000 * t]`, pad to multiple of frame size (1920), run seanet encoder to downsample to `[200 * t, 512]`
    - Run encoder transformer on the `[200 * t, 512]`
    - Downsample with strided convolution to `[12.5 * t, 512]`
  - Apply `speaker_proj_weight` dense layer to expand into `[12.5 * t, 1024]`
- **Text conditioning `LUTConditioner` (what text for the model to "speak")**
  - Apply SentencePiece encoder with Unigram mode, produce raw list of tokens
  - Lookup tokens in a learned embedding matrix `[4001, 1024]` — vocab size (`n_bins` in code) is 4000
- **Sampling (repeat for each frame)**
  - Concatenate text embeddings with voice embedding (both dim 1024), and also concatenate input projection (512->1024) applied to previous audio latents (dim 512)
  - Run backbone transformer to sample next audio token in `[S, 1024]` -> `[S+1, 1024]`
  - Audio latent sampling via diffusion FlowNet `SimpleMLPAdaLN`
    - Sample isotropic Gaussian noise `x` with Var[X]=temp, and use a 1-step integrator, trained with "Lagrangian Self Distillation"
    - For the integrator, project transformer output via `cond_embed` into the latent dim `[512]`, add time embedding to produce `y`
    - Run depth-6 residual MLP passing `x` along, with `y` used to modulate AdaLN and SwiGLU gates -> final projection to `[32]`
  - Ok now we have the audio latent `[32]`, decode it with KV cache with (Mimi decode)
    - Run bottleneck through "dummy quantizer" from `[32]` to `[512]`
    - Upsample with transposed depthwise convolution from `[1, 512]` (note: 12.5 Hz frame) to `[16, 512]` (200 Hz latent)
    - Run the decoder transformer on `[16, 512]`
    - Finally, run seanet decoder to upsample to `[1920]` audio samples

This is pretty complicated, let's think about the right way to structure this. I'll probably start without aiming to JIT things and just create each building block separately. Also no need for KV cache until we get inference _correct_ first.

**Comment by gabrieldemarmiesse** [comment]
Another trick we use to lower the latency is that we cache the model state after the voice embeddings (conditioning) has been fed to the model, meaning we can easily clone this state to do a new generation without having to re-feed the audio embeedings as prefix to the model.

See the function `get_state_for_audio_prompt`

**Comment by ekzhang** [comment]
Today's progress:

- Added some library functions `jax.nn.dotProductAttention()` and `jax.lax.convTranspose()` that will be useful for this model
  - Also they're just generally pretty useful, I did these using lower-level APIs in other examples but it's useful to have the real APIs
- Loaded the pocket tts model weights in browser

<img width="886" height="539" alt="Image" src="https://github.com/user-attachments/assets/811d6fff-b63d-4478-be5d-2b7cc2338620" />

**Comment by ekzhang** [comment]
Another progress update:

- SentencePiece tokenizer is working and tested (admittedly this was vibe-coded, but it seems to be correct)
- FlowLM inference is built, and I tested it against a sample output with Azelma, outputs match Python implementation closely (although not exact because of the use of float16 instead of bfloat16 quant)
  - Layer norm and RMS Norm are running in float32 still, don't worry about the numerics of that haha
- Going to test Mimi next

After getting a basic inference demo working, I'll implement caching (kv and stateful convolution) and then ship it

**Comment by ekzhang** [comment]
https://github.com/user-attachments/assets/71c01211-340c-4b92-9d51-38dcedff9531

A bit faster, no kv cache yet so it's doing ~256x more flops than needed
